<a href="https://colab.research.google.com/github/MarkNgendo/MTN-Cote-d-Ivoire-Project/blob/main/Moringa_Data_Science_Prep_W3_Independent_Project_2021_03_Mark_Ngendo_NoteBook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **WEEK THREE INDEPENDENT PROJECT PYTHON**

## **PROBLEM STATEMENT**


Currently MTN Cote d'Ivoire would like to upgrade its technology infrastructure for its mobile users in Ivory Coast. Studying the given dataset, how does MTN Cote d'Ivoire go about the upgrade of its infrastructure strategy within the given cities?

## **DATASETS PROVIDED**
Below is a list of datasets that we shall be working on:


*  cells_geo_description.xlsx
*   cells_geo.csv 
*   CDR_description.xlsx 
*   CDR 20120507 
*   CDR 20120508 
*   CDR 20120509 










## Importing python libraries and loading datasets.





In [69]:
import pandas as pd
import numpy as np

In [70]:
cells = pd.read_csv('cells_geo.csv', delimiter=';')
cells.head()

Unnamed: 0   VILLES      STATUS  ... AREA     CELL_ID   SITE_CODE
0           0   ADJAME  In Service  ...  NaN  fbb2e75af3  0bb492ef17
1           1   ADJAME  In Service  ...  NaN  1502501d78  0bb492ef17
2           2   ADJAME  In Service  ...  NaN  0c2e182a0c  0bb492ef17
3           3  ASSINIE  In Service  ...  NaN  69bdfc69fd  8a1e6ce31f
4           4  ASSINIE  In Service  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 12 columns]

In [71]:
geo = pd.read_excel('cells_geo_description.xlsx')
geo

Column name                                        Description  Format
0         VILLES                                               City  String
1         STATUS                                  In Service or not  String
2   LOCALISATION                                  in ABIDJAN or not  String
3     DECOUPZONE                                  Geographical Zone  String
4       ZONENAME                                       Name of Zone  String
5      LONGITUDE                                          Longitude   Float
6       LATITUDE                                           Latitude   Float
7         REGION                                             Region  String
8           AREA                                               Area  String
9        CELL_ID                                     ID of the cell  String
10     SITE_CODE  Site (there are several cells per site, severa...  String

In [72]:
cdr = pd.read_excel('CDR_description.xlsx')
cdr

Column name                                        Description   Format
0          PRODUCT                                       Voice or SMS   String
1            VALUE                                      Billing price  Integer
2        DATE_TIME               Time in format yyyy-MM-dd hh:mm:ss.0   String
3     CELL_ON_SITE  Which cell in the site was used (not needed here)  Integer
4  DW_A_NUMBER_INT  Anonymized phone number of the person for whic...   String
5  DW_B_NUMBER_INT        Anonymized phone number of the counterparty   String
6        COUNTRY_A                  Country of party A (useless here)   String
7        COUNTRY_B                  Country of party B (useless here)   String
8          CELL_ID                                     ID of the cell   String
9          SITE_ID                                     ID of the SITE   String

In [73]:
set1 = pd.read_csv('Telcom_dataset.csv')
set1.head()

PRODUTC  VALUE               DATETIME  ... COUNTRY_B     CELL_ID     SITE_ID
0   Voice      0  2012-05-06 23:04:37.0  ...     19e2e  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  2012-05-06 23:05:25.0  ...     19e2e  /+cKZKIp41  /+cKZKIp41
2     sms     25  2012-05-06 23:05:41.0  ...     19e2e  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  2012-05-06 23:05:42.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  2012-05-06 23:06:29.0  ...     19e2e  051KhYDCpv  051KhYDCpv

[5 rows x 10 columns]

In [74]:
set2 = pd.read_csv('Telcom_dataset2.csv')
set2.head()


PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B     CELL_ID     SITE_ID
0     sms      0  2012-05-07 23:02:06.0  ...     19e2e  +854AcBQT2  +854AcBQT2
1   Voice      0  2012-05-07 23:03:44.0  ...     19e2e  +854AcBQT2  +854AcBQT2
2     sms      0  2012-05-07 23:04:06.0  ...     19e2e  +laSrk7g6q  +laSrk7g6q
3   Voice     37  2012-05-07 23:04:48.0  ...     19e2e  /xvRrCVKoQ  /xvRrCVKoQ
4     sms     25  2012-05-07 23:04:58.0  ...     19e2e  +bO+qkSonO  +bO+qkSonO

[5 rows x 10 columns]

In [75]:
set3 = pd.read_csv('Telcom_dataset3.csv')
set3 = set3.rename(columns= {"SIET_ID": "SITE_ID"})
set3.head()

PRODUCT  VALUE              DATE_TIME  ... COUNTRY_B      CELLID     SITE_ID
0   Voice     61  2012-05-08 23:01:28.0  ...     19e2e  +38u2u/rfx  +38u2u/rfx
1     sms      0  2012-05-08 23:01:31.0  ...     19e2e  +Fs4mTvdKx  +Fs4mTvdKx
2   Voice     10  2012-05-08 23:02:59.0  ...     19e2e  +ZaaZozy/+  +ZaaZozy/+
3     sms      0  2012-05-08 23:03:04.0  ...     19e2e  +cNeJzsTp3  +cNeJzsTp3
4   Voice     22  2012-05-08 23:03:16.0  ...     19e2e  /eeMklgyJA  /eeMklgyJA

[5 rows x 10 columns]

The next step involves data exploration before cleaning and eventual analysis can be done.

### Looking for ways to make the data more readable.

In the cells dataframe, the 'Abidjan_EST' entry has unintended quotations from the source. Also, since each Telecom dataset represent specific dates, the date-time column needs to only represent time by itself. As such, it needs to be split and renamed. In the cells df, the Unamed column needs to be dropped as it serves no purpose.



##DATA PREPARATION AND CLEANING

### Editing the dataframe column names

In [76]:
cells = cells.rename(columns={'VILLES': 'City'})
cells = cells.replace(to_replace='"Abidjan_EST', value='Abidjan_EST')
del cells['Unnamed: 0']
cells.head()

City      STATUS LOCALISATION  ... AREA     CELL_ID   SITE_CODE
0   ADJAME  In Service      ABIDJAN  ...  NaN  fbb2e75af3  0bb492ef17
1   ADJAME  In Service      ABIDJAN  ...  NaN  1502501d78  0bb492ef17
2   ADJAME  In Service      ABIDJAN  ...  NaN  0c2e182a0c  0bb492ef17
3  ASSINIE  In Service    INTERIEUR  ...  NaN  69bdfc69fd  8a1e6ce31f
4  ASSINIE  In Service    INTERIEUR  ...  NaN  7a1029d2ab  8a1e6ce31f

[5 rows x 11 columns]

In [77]:
set1 = set1.rename(columns={"PRODUTC": "PRODUCT"})
set1['VALUE'].groupby(set1['PRODUCT']).sum()

PRODUCT
Voice    108869
data       7032
sms       13473
Name: VALUE, dtype: int64

### Eliminating useless columns in each dataset

In [78]:
eliminate = ['CELL_ON_SITE', 'COUNTRY_A', 'COUNTRY_B']
set1.drop(eliminate, axis=1,inplace=True)
set2.drop(eliminate, axis=1, inplace=True)
set3.drop(eliminate, axis=1, inplace=True)

### Checking and removing nulls

In [79]:
cells.isnull().sum().sort_values(ascending=False)

STATUS          67
AREA            23
ZONENAME         6
SITE_CODE        0
CELL_ID          0
REGION           0
LATITUDE         0
LONGITUDE        0
DECOUPZONE       0
LOCALISATION     0
City             0
dtype: int64

In [112]:
set1.isnull().sum().sort_values(ascending=False)

SITE_ID            704
CELL_ID              0
DW_B_NUMBER_INT      0
DW_A_NUMBER_INT      0
DATETIME             0
VALUE                0
PRODUCT              0
dtype: int64

In [81]:
set2.isnull().sum().sort_values(ascending=False)

SITE_ID        676
CELL_ID          0
DW_B_NUMBER      0
DW_A_NUMBER      0
DATE_TIME        0
VALUE            0
PRODUCT          0
dtype: int64

In [82]:
set3.isnull().sum().sort_values(ascending=False)

SITE_ID            605
CELLID               0
DW_B_NUMBER_INT      0
DW_A_NUMBER_INT      0
DATE_TIME            0
VALUE                0
PRODUCT              0
dtype: int64

### checking for duplicated data in the days sets before merging them

In [113]:
cells.duplicated().value_counts()

False    3931
dtype: int64

In [83]:
set1.duplicated().value_counts()

False    4919
True       82
dtype: int64

In [84]:
set2.duplicated().value_counts()

False    4923
True       78
dtype: int64

In [85]:
set3.duplicated().value_counts()

False    4908
True       93
dtype: int64

In [86]:
cells.drop_duplicates(inplace=True)
set1.drop_duplicates(inplace=True)
set2.drop_duplicates(inplace=True)
set3.drop_duplicates(inplace=True)

In [93]:
# Merging the three days' datasetes
days = pd.concat([set1,set2,set3], axis=0, join='outer', ignore_index=True)
days.head()

PRODUCT  VALUE               DATETIME  ... DW_A_NUMBER DW_B_NUMBER CELLID
0   Voice      0  2012-05-06 23:04:37.0  ...         NaN         NaN    NaN
1     sms      0  2012-05-06 23:05:25.0  ...         NaN         NaN    NaN
2     sms     25  2012-05-06 23:05:41.0  ...         NaN         NaN    NaN
3   Voice     56  2012-05-06 23:05:42.0  ...         NaN         NaN    NaN
4   Voice     14  2012-05-06 23:06:29.0  ...         NaN         NaN    NaN

[5 rows x 11 columns]

### Enhancing the new concatated table

In [94]:
del days['DATE_TIME']
del days['CELLID']
del days['DW_A_NUMBER']
del days['DW_B_NUMBER']
days.head()

PRODUCT  VALUE  ...     CELL_ID     SITE_ID
0   Voice      0  ...  /TJNe+Mmtu  /TJNe+Mmtu
1     sms      0  ...  /+cKZKIp41  /+cKZKIp41
2     sms     25  ...  /u0FSD+ahi  /u0FSD+ahi
3   Voice     56  ...  +cNeJzsTp3  +cNeJzsTp3
4   Voice     14  ...  051KhYDCpv  051KhYDCpv

[5 rows x 7 columns]

###Merging all datasets now and cleaning them

In [95]:
total_df = pd.merge(days, cells, how='inner', left_on=['SITE_ID'], right_on=['SITE_CODE'])
total_df.head()

PRODUCT  VALUE               DATETIME  ...          AREA   CELL_ID_y   SITE_CODE
0     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  24ccdda3d5  1b5540c02d
1     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ebfc58af70  1b5540c02d
2     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  ba8ce6e4ef  1b5540c02d
3     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  c2545171dc  1b5540c02d
4     sms      0  2012-05-07 00:00:00.0  ...  CENTRE NORTH  f15bd5483f  1b5540c02d

[5 rows x 18 columns]

In [96]:
eliminate = ['DW_B_NUMBER_INT', 'CELL_ID_x', 'CELL_ID_y', 'SITE_CODE']
total_df.drop(eliminate, axis=1, inplace=True)

## ANALYSIS

In [97]:
#Regions with the highest number of calls.
total_df['REGION'].value_counts()

LAGUNE               32547
LACS                  2408
BAS SASSANDRA         2327
HAUT SASSANDRA        2277
VALLEE DU BANDAMA     1551
SUD BANDAMA           1518
SAVANE                 980
FROMAGER               904
WORODOUGOU             792
18 MONTAGNES           765
MARAHOUE               618
AGNEBY                 581
NZI COMOE              564
MOYEN CAVALLY          528
DENGUELE               395
ZANZAN                 330
MOYEN COMOE            315
SUD COMOE              203
BAFING                  99
Name: REGION, dtype: int64

Taking the top 10 regions with highest traffic:
* LAGUNE               32547
* LACS                  2408
* BAS SASSANDRA         2327
* HAUT SASSANDRA        2277
* VALLEE DU BANDAMA     1551
* SUD BANDAMA           1518
* SAVANE                 980
* FROMAGER               904
* WORODOUGOU             792
* 18 MONTAGNES           765


In [104]:
high_regions =['LAGUNE', 'LACS', 'BAS SASSANDRA', 'HAUT SASSANDRA', 'VALLEE DU BANDAMA',
              'SUD BANDAMA', 'SAVANE', 'FROMAGER' 'WORODOUGOU' ,'MONTAGNES']

hr_df = total_df[total_df['REGION'].isin(high_regions)]
hr_df.head()

PRODUCT  VALUE  ...             REGION          AREA
0     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
1     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
2     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
3     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
4     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH

[5 rows x 14 columns]

In [101]:
#Which Cities are in high traffic zones
total_df['ZONENAME'].value_counts()

ZONE C - Abidjan_OUEST     10296
ZONE E - Abidjan_NORD       6660
ZONE B - Grand-SUD          5545
ZONE B - Abidjan_EST        4838
ZONE D - Abidjan_SUD        4642
ZONE D - Grand-OUEST        4314
ZONE A - Abidjan_CENTRE     4132
ZONE A - Grand-NORD         3928
ZONE E - SUD-OUEST          2993
ZONE C - Grand-EST          1748
ZONE D - Abidjan_EST         186
OKAHOTEL                     120
NPONON                        72
WILLIAMSVILLE2                40
ZONE C - Abidjan_EST          27
PALMABINGERVILLE              21
ZONE A - Abidjan_OUEST        21
ZONE B - Grand-EST            18
ZONE A - Grand-SUD            18
ZONE B - Abidjan_OUEST        18
ZONE C - Abidjan_SUD          15
ZONE B - Abidjan_SUD           6
ZONE D - Grand-SUD             6
ZONE A - Grand-EST             6
ZONE B - Abidjan-EST           3
ZONE B - Grand-NORD            3
ZONE B - Abidjan_Nord          3
AGBANNEW                       3
ZONE B - Abidjan_CENTRE        2
Name: ZONENAME, dtype: int64

The top 10 cities in the high traffic zones are:
* ZONE C - Abidjan_OUEST     10296
* ZONE E - Abidjan_NORD       6660
* ZONE B - Grand-SUD          5545
* ZONE B - Abidjan_EST        4838
* ZONE D - Abidjan_SUD        4642
* ZONE D - Grand-OUEST        4314
* ZONE A - Abidjan_CENTRE     4132
* ZONE A - Grand-NORD         3928
* ZONE E - SUD-OUEST          2993
* ZONE C - Grand-EST          1748

In [103]:
high_zones = ['ZONE C - Abidjan_OUEST','ZONE E - Abidjan_NORD', 'ZONE B - Grand-SUD',
              'ZONE B - Abidjan_EST', 'ZONE D - Abidjan_SUD', 'ZONE A - Grand-NORD',
              'ZONE E - SUD-OUEST', 'ZONE C - Grand-EST', 'ZONE D - Abidjan_EST']
hz_df = total_df[total_df['ZONENAME'].isin(high_zones)]
hz_df.head()

PRODUCT  VALUE  ...             REGION          AREA
0     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
1     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
2     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
3     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH
4     sms      0  ...  VALLEE DU BANDAMA  CENTRE NORTH

[5 rows x 14 columns]

In [105]:
# What is product usage like in these high traffic zones?
usage = hz_df['PRODUCT'].value_counts()
usage

sms      23820
Voice    17016
Name: PRODUCT, dtype: int64

In [108]:
#Which of the two is the more profitable product?
profit = hz_df.groupby(['PRODUCT'])['VALUE'].sum()
(profit / profit.sum()) * 100

PRODUCT
Voice    83.264754
sms      16.735246
Name: VALUE, dtype: float64

In [111]:
# Most profitable zone is?
pd.pivot_table(hz_df, values='VALUE',index= ['PRODUCT','ZONENAME'], aggfunc=np.sum)

VALUE
PRODUCT ZONENAME                      
Voice   ZONE A - Grand-NORD     102150
        ZONE B - Abidjan_EST    183981
        ZONE B - Grand-SUD       99989
        ZONE C - Abidjan_OUEST  173149
        ZONE C - Grand-EST       78881
        ZONE D - Abidjan_EST      8196
        ZONE D - Abidjan_SUD     99990
        ZONE E - Abidjan_NORD   108916
        ZONE E - SUD-OUEST       84489
sms     ZONE A - Grand-NORD      13198
        ZONE B - Abidjan_EST     25480
        ZONE B - Grand-SUD       23634
        ZONE C - Abidjan_OUEST   52278
        ZONE C - Grand-EST        7643
        ZONE D - Abidjan_EST       750
        ZONE D - Abidjan_SUD     24562
        ZONE E - Abidjan_NORD    31101
        ZONE E - SUD-OUEST       10231